### ISSUE TITLE
tfp.stats.histogram does not support negative values

### ISSUE THREAD

**Issue by dmadisetti** [tensorflow/probability#1430]
This is due to `tf.math.bincount` not supporting negative values.

There's a pretty trivial fix which is just offsetting the value of bins to be non-negative and then them shifting back: https://github.com/tensorflow/probability/blob/b8ac98db180896dd938cf171b7b22a06e21e02ac/tensorflow_probability/python/stats/quantiles.py#L414
Before I implement this fix, I just wanted a sanity check, does this sound right?

Related: b/124015136

reproducing example:

```python
x = tf.stack([tf.random.normal([1000]), tf.random.normal([1000])],
             axis=-1)

edges = [0., 0.5, 1.0, 1.5, 2.0]

tfp.stats.histogram(x, edges)
```

<details>
  <summary>Stacktrace</summary>
  
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-35-2be6b2b6813e> in <module>()
      4 edges = [0., 0.5, 1.0, 1.5, 2.0]
      5 
----> 6 tfp.stats.histogram(x, edges)
      7 
      8 tfp.stats.histogram(x, edges, axis=0)

5 frames
/usr/local/lib/python3.7/dist-packages/tensorflow_probability/python/stats/quantiles.py in histogram(x, edges, axis, weights, extend_lower_interval, extend_upper_interval, dtype, name)
    422         maxlength=tf.shape(edges)[0] - 1,
    423         axis=0,
--> 424         dtype=dtype or in_dtype)
    425     n_edges = tf.compat.dimension_value(edges.shape[0])
    426     if n_edges is not None:

/usr/local/lib/python3.7/dist-packages/tensorflow_probability/python/stats/quantiles.py in count_integers(arr, weights, minlength, maxlength, axis, dtype, name)
    104           minlength=minlength,
    105           maxlength=maxlength,
--> 106           dtype=dtype)
    107 
    108     # Move dims in ~axis to the left, so we can tf.map_fn bincount over them,

/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/bincount_ops.py in bincount(arr, weights, minlength, maxlength, dtype, name, axis, binary_output)
    139         return gen_math_ops.unsorted_segment_sum(weights, arr, output_size)
    140       weights = constant_op.constant([], dtype)
--> 141       return gen_math_ops.bincount(arr, output_size, weights)
    142 
    143     if not isinstance(arr, sparse_tensor.SparseTensor):

/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/gen_math_ops.py in bincount(arr, size, weights, name)
   1804       return _result
   1805     except _core._NotOkStatusException as e:
-> 1806       _ops.raise_from_not_ok_status(e, name)
   1807     except _core._FallbackException:
   1808       pass

/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py in raise_from_not_ok_status(e, name)
   6939   message = e.message + (" name: " + name if name is not None else "")
   6940   # pylint: disable=protected-access
-> 6941   six.raise_from(core._status_to_exception(e.code, message), None)
   6942   # pylint: enable=protected-access
   6943 

/usr/local/lib/python3.7/dist-packages/six.py in raise_from(value, from_value)

InvalidArgumentError: Input arr must be non-negative! [Op:Bincount]
</details>

**Comment by dmadisetti** [comment]
Hmm. `probability`'s histogram may be over-engineered, it looks like there's builtin `tf.histogram_fixed_width_bins` and `tf.histogram_fixed_width`

I am no longer interested in patching this, [I found a more robust solution](https://gist.github.com/dmadisetti/ea7b46aa0457d757a4b52057b5b9f5ea), but I do think this is still and issue, so I'll leave it open

**Comment by jonas-eschle** [comment]
Negative values work actually with histogram. However, in the example that you've shown, there are values that go beyond the edges and it's not specified that the lower edge should be extended.

I agree that the error could be more verbose.

The example will work if we create edges that contain all the values:
```
x = tf.stack([tf.random.normal([1000]), tf.random.normal([1000])],
             axis=-1)

edges = [-10, 0., 0.5, 1.0, 1.5, 2.0, 10]

tfp.stats.histogram(x, edges)
```

or by having an overflow bin
```
x = tf.stack([tf.random.normal([1000]), tf.random.normal([1000])],
             axis=-1)

edges = [0., 0.5, 1.0, 1.5, 2.0]

tfp.stats.histogram(x, edges, extend_lower_interval=True, extend_upper_interval=True)
```

Are you interested to work with histograms? What is your use-case (out of interest)?

**Comment by dmadisetti** [comment]
Thanks for the response! I guess I should have dug a bit deeper into the api- but then the behavior is inconsistent between boundaries:

```python
import tensorflow as tf
import tensorflow_probability as tfp

x = tf.stack([tf.random.normal([1000]), tf.random.normal([1000])],
             axis=-1) + 10

edges = [0., 0.5, 1.0, 1.5, 2.0]

tfp.stats.histogram(x, edges)
```

This runs and produces 0s.

Re use-case:
My plan is to cast higher dimension data to a plane, and then use histogram to "pixelize" the result. I then use this pixelized result in my pipeline. Ideally going through `tf` so that I can backprop to those higher dimensions. I will likely just use the gist I provided, since it removes another dependency.

**Comment by jonas-eschle** [comment]
Yes, I agree that the behavior is inconsistent. Maybe change the title to this, then it can be kept open
